### ISSUE TITLE
Lists are slow; why not using hash tables?

### ISSUE THREAD

**Issue by jcguu95** [org-roam/org-roam#2241]
### Brief Abstract
I have many nodes and things start getting slower and slower in org-roam. I suspect one core issue is that it uses lists rather than hash table (e.g. `#'org-roam-node-list`). **I want to help.**

### Long Description
I can count myself as a "long-term" org-roam user. Thanks for the wonderful package! However, that means I have thousands of nodes, and each time the db update itself, it takes more than 20 seconds. 

I have a tiny hack uses memoization. However, it takes a lot of memory and sometimes it still gets slow.
``` emacs-lisp
  (setq org-roam-node-default-sort nil)
  (require 'memoize)
  (memoize 'org-roam-node-read--completions "10 minute")
  (defun my/force-update-org-roam-node-read (&optional timeout)
    (interactive)
    (flet ((memoize-force-update (fn-symbol &optional timeout)
                                 (when (get fn-symbol :memoize-original-function)
                                   (progn (memoize-restore fn-symbol)
                                          (memoize fn-symbol timeout)))))
      (memoize-force-update 'org-roam-node-read--completions
                            (or timeout memoize-default-timeout))))
  (run-with-idle-timer 60 t #'my/force-update-org-roam-node-read)
```

**QUESTION** Why do we use lists instead of hash tables? With the latter, it's easier to update and make changes to a specific node without traversing over the whole thing.

### Proposed Implementation (if any)
I'd like to know the principle of design behind, and how much it would need to change. And then I will come up with a concrete implementation.

**Comment by jethrokuan** [comment]
That's primarily because the emacs APIs (e.g. minibuffer completion) take in a list and not a hash table. Traditionally hash tables have not been very popular and association lists are also used in place of hash tables.

**Comment by jcguu95** [comment]
Indeed, the traditional `#'completing-read` only takes lists. However, nowadays many people have hooked it so it calls some modern completion framework instead. For example, `#'ivy-read` and `#'consult--read` both take hash tables.

A proposal would be to let the user choose if they want to enable hash table support. In that case, **(1)** do you think this will make org-roam (with thousands of nodes) significantly faster? For example, we can update the very specific entry of the database only when we need to, without touching (e.g. traversing through the list) other nodes at all. **(2)** What are some starting points and caveats that you'd advice me to pay attention to? We don't need to make this into the official branch, but I think it will be a nice thing to have, especially for those who have thousands of nodes like me :)

```
(length (org-roam-node-list)) ;; => 5704      ; :D :D
```

**Comment by hwiorn** [comment]
Recently, I've felt `org-roam-find-node` is really slow. It takes over 4~5 seconds by default when I search and insert notes every time. `org-roam-list` takes 0.5 seconds itself, and it's not slow. The templates make `consult` slower.  I think org-roam performance can be matter in the near future.

But My nodes are
```
(length (org-roam-node-list)) ;; => 8594
```

I use `hyp.is to org` converter my own version. You can see [karlicoss/orger](https://github.com/karlicoss/orger) and [karlicoss/hypexport](https://github.com/karlicoss/hypexport).
I translate all highlight memos to headlines which has ID. It makes searchable notes and takes easily from literature(hyp.is) to my permanent notes.

I tried `memoize` of `org-roam-find-node` but it didn't work.
Your code helped me a lot. Thank you for your code, @jcguu95

I think it didn't solve the performance issuse, But I can focus my note-taking now.

**Comment by jcguu95** [comment]
@hwiorn Glad you found it helpful.

@jethrokuan I just checked the doc of `#'completing-read` in emacs `28.2`, which mentions that the `COLLECTION` parameter of `#'completing-read` can take a hash table. Am I missing anything?

> COLLECTION can be a list of strings, an alist, an obarray or a hash table. COLLECTION can also be a function to do the completion itself.

```
completing-read is a function defined in C source code.

Signature
(completing-read PROMPT COLLECTION &optional PREDICATE REQUIRE-MATCH INITIAL-INPUT HIST DEF INHERIT-INPUT-METHOD)

Documentation
Read a string in the minibuffer, with completion.

PROMPT is a string to prompt with; normally it ends in a colon and a space.
COLLECTION can be a list of strings, an alist, an obarray or a hash table.
COLLECTION can also be a function to do the completion itself.
PREDICATE limits completion to a subset of COLLECTION.
See try-completion, all-completions, test-completion,
and completion-boundaries, for more details on completion,
COLLECTION, and PREDICATE.  See also Info node (elisp)Basic Completion
for the details about completion, and Info node (elisp)Programmed
Completion for expectations from COLLECTION when it's a function.
```

**Comment by hpfr** [comment]
Looks like completing-read has taken hash tables for many years
